### DISCUSSION TITLE
如何在子工程里修改全局编译选项

### DISCUSSION THREAD

**Discussion by Jiu-xiao** [xmake-io/xmake#2760] [Q&A] [answered]
我正在将一个带有bsp管理的cmake工程移植到xmake，工程结构为
```
├── bsp
│   ├── board1
│   │   └── xmake.lua
│   └── board2
│       └── xmake.lua
├── src
│   └── xmake.lua
└── xmake.lua

```
bsp内定义了不同开发板的编译器，宏定义，链接脚本和C_FLAG等参数，我无法通过include某个board的xmake.lua来覆盖全局的编译参数。请问是否有某种方法通过xmake实现bsp管理，选择某个子项目的配置并将其应用到其他所有子项目

**Comment by waruqi** [comment, marked as answer]
用 set_xxx 不要用 add_xxx 就是覆盖，另外你在子 board/xmake.lua 的 root 域设置，就是局部全局配置

**Reply by Jiu-xiao** [reply]
如果在./xmake.lua中include ./bsp/board1/xmake.lua和./src/xmake.lua，在./bsp/board1/xmake.lua的根域中的配置并不能作用到./src/xmake.lua中。我尝试了另一种写法
./bsp/board1/xmake.lua
```
target("board1")
add_defines("DEBUG", {public = true})
......
```
./src/xmake.lua
```
target("src")
add_deps("board1")
```
这样只要在级联的的依赖中存在board1就可以继承它的配置，又因为其他子目标都直接或间接的依赖了bsp，所以我可以将board1的配置应用到全局。是否还有更好的实现方法？

**Reply by waruqi** [reply]
> 如果在./xmake.lua中include ./bsp/board1/xmake.lua和./src/xmake.lua，在./bsp/board1/xmake.lua的根域中的配置并不能作用到./src/xmake.lua中。我尝试了另一种写法

根域配置按 tree 状向下继承，src/xmake.lua 的配置，bsp/board1/xmake.lua 里面能继承到，也能局部覆盖，但不能反过来影响上层配置，设计就是如此。

你可以 src/xmake.lua 根域放置通用配置，bsp/boardX/xmake.lua 不同子工程放置 board 仅有的配置。

当然走 add_deps/public 依赖继承也可以，主要是配置存在链接依赖 target 间的配置传递，看你自己项目的需求，这两者应用场景不同
